### ISSUE TITLE
Test keyboard navigation for drag and drop interfaces

### ISSUE THREAD

**Issue by kimisgold** [omeka/Omeka#1098]
This is a feature that addresses #1035, providing a keyboard navigation alternative to parts of the admin that use drag and drop patterns for manipulating trees. On the [`keyboard-drag-and-drop`](https://github.com/omeka/Omeka/tree/keyboard-drag-and-drop) branch, the current solution is implemented on the view for editing the site navigation (Appearance > Navigation). The functionality includes the following:
* Each navigation item should know have a keyboard button that opens a group of buttons for reordering. 
  * The user can move the navigation item up and down the tree. 
  * They can also nest the navigation item under a preceding node, or unnest from a parent node. 
  * After each reorder action, there should be a screen reader alert.
    * If the action fails, i.e. it can go no higher in the tree or cannot be nested further, there should be a general failure alert saying "Cannot reorder further."
    * If the action succeeds, it should announce 1) the navigation item reordered, and 2) the direction of reordering and most relevant node. For example: "Browse Items reordered. Moved above About."
* Activating the keyboard button should close the group of reorder actions and announce itself collapsed.

Test across desktop and mobile browsers with the usual suite of screen readers. Beyond feedback for testing functionality, this is also a good place for suggesting expanded features and improvements.

**Comment by aviannamiller** [comment]
This is a really great addition to provide equivalent functionality. Great work on this! Some preliminary feedback:
- With longer page labels that overflow onto another line, the keyboard reorder button and drawer toggle buttons are pushed slightly to the right, out of alignment with the rest, and the toggle is cut off:
  <img width="411" height="288" alt="Four links in a navigation tree, all accompanied by a keyboard icon button and a toggle button. One of the link labels is longer, and its accompanying buttons are pushed slightly to the right" src="https://github.com/user-attachments/assets/667cee1c-e080-4b66-a6a3-d4229ff6b6cb" />
    - The toggle is also then almost imperceivable when the keyboard reorder panel is open:
      <img width="424" height="104" alt="Open keyboard reorder panel in a drawer. Only a small edge of the toggle icon is visible as it is pushed all the way to the right." src="https://github.com/user-attachments/assets/2c4b9995-6ccc-4745-a4ba-a282a49eb4e0" />
- This feature is not working on links added directly in to navigation on this page. Though the keyboard reorder button appears, the panel does not open, unless it is marked for deletion. Then, the functionality returns, and the drawer can be moved while it is marked or after it is then subsequently unmarked.
    - As a tangential point, I’m not sure there is a need for the keyboard reordering features to remain on the drawer once its marked for deletion, although the drawer can technically still be dragged in this state, so it does provide equivalent functionality to have it… If the buttons will remain, the spacing between them and the undo button should be adjusted, both for the keyboard button alone and the open panel:
      <img width="84" height="113" alt="Keyboard reorder icon next to a toggle button, compared to the keyboard icon next to an undo icon. There is much less spacing between the two in the latter example." src="https://github.com/user-attachments/assets/660782ae-4e9b-4848-bc6d-65f15a917ea3" />
      <img width="171" height="62" alt="Open keyboard reorder panel. Undo icon is placed very closely to the panel buttons." src="https://github.com/user-attachments/assets/d9eab743-694f-4f8c-bd1c-b605794fb832" />

I'm still testing, especially across different platforms and screen readers, so full results still to come.

**Comment by aviannamiller** [comment]
Remaining results:

- I am not getting consistent error alerts with JAWS on Chrome and Edge and VoiceOver on Chrome, Safari, and Firefox. I am noticing this sometimes when the item is in a state where no further actions are available (i.e. attempting to unnest or move upward the most top-level link), or in cases where an impossible action, so to speak, is attempted, in which case an alert may be present, but then selecting another option that is also not possible is then not accompanied by the “Cannot reorder further” alert. Again, this is not always the case, but the alerts are just generally inconsistent for these screen reader / browser combinations. On the flip side, NVDA and JAWS/Firefox consistently notify of the alerts.
- VoiceOver/Firefox is not consistently announcing directional information in relation to other nodes. I’m getting “Unnested from.”/“Nested from.” without the name of the relevant node in there on the first try nesting/unnesting, and subsequent nesting/unnesting actions result in just the name of the associated node, (i.e. just “About,” with no “Unnested from”/“Nested from” attached to it).
- NVDA on Edge, Firefox, and Chrome is not announcing anything at all about position in relation to nodes or direction of reordering.
- With NVDA for both Chrome and Edge, I’m getting some strange jumping around of focus when attempting to unnest items:

https://github.com/user-attachments/assets/28b0f5f4-8e8f-4a77-ba71-23e8bf1c835d

Video description: When attempting to unnest the "More Information" item, focus jumps to the previous focusable element (outside of the keyboard reorder panel) after a consecutive selection of the unnest button, in this case, the drawer toggle. When returning to the unnest button and attempting to activate it again, focus then jumps backwards to the now visible public page link.

**Comment by kimisgold** [comment]
Newly added nav items should now reorder properly. The keyboard reordering should also fully deactivate when marking a nav item for deletion.

I'm still working on better layout for long titles and narrower device widths, but I'd like to get those other fixes looked at in the meantime.

**Comment by aviannamiller** [comment]
Great, the pattern is working well for newly added nav items, for both facets of the fixes. Perhaps a bit outside the issue, but once the new nav item moves up, focus becomes a bit lost. Progressing brings you to "Save Changes", but I would think it more appropriate to direct to the keyboard reorder button on the item (or its remove button, potentially).

**Comment by kimisgold** [comment]
**Long nav item labels:** I've constrained nav items to a single line with an ellipsis indicating truncation. It's not ideal, but if the user needs to see the longer label, the full label is accessible when expanding the nav drawer.

**Focus after adding a new nav item:** Thinking on it some, I've redirected focus after creating a nav item to the label input on the "Add a Link to the Navigation" panel, so the user can quickly begin working on a new nav item. Is that too big an assumption?

**Other drag and drop interfaces:** I've generalized the JS for the keyboard sorting and applied it to the rest of the admin. They're ready for testing on the following interfaces:

* Element Sets > Edit
* Item > Files
* Item Types > Edit

**Comment by kimisgold** [comment]
Just pushed a fix for the screen reader announcements: they should be consistently announcing the titles of relevant nodes when performing a reorder action.

**Comment by aviannamiller** [comment]
> **Long nav item labels:** I've constrained nav items to a single line with an ellipsis indicating truncation. It's not ideal, but if the user needs to see the longer label, the full label is accessible when expanding the nav drawer.

This seems like a good solution. Can the right margin be cut down at all? It feels like it's cutting them off with quite a bit of space left to go, but perhaps there's a practical reason for this. Example:
<img width="565" height="97" alt="Nav item drawer with long name truncated with ellipsis. Title is selected via dev tools, revealing large right side margin." src="https://github.com/user-attachments/assets/f15fdec7-9638-4a64-83af-7e7d75f63302" />

> **Focus after adding a new nav item:** Thinking on it some, I've redirected focus after creating a nav item to the label input on the "Add a Link to the Navigation" panel, so the user can quickly begin working on a new nav item. Is that too big an assumption?

This feels like a good workflow, for now. Further feedback could certainly be gathered on this, as with similar patterns across the interface. Can an announcement be added for after the link is added to indicate success and that focus is now within a new field to allow for adding an additional link? Something like, "Link added," followed by label info of the new input focus has moved to.

**Other interfaces:**
- For all areas where the pattern has been newly added, draggable handles can be navigated to via keyboard. These elements had been made unfocusable in nav appearance, which seems like the right move since they are redundant and not functional via this mode.
- For all implementations, JAWS/Firefox is still not announcing locations, just control labels.
- NVDA directional announcements are now coming through properly, except where noted below.
- **Item Types:**
  - Focus keeps appearing to become stuck on buttons right above or below move up/move down buttons (while solely and consecutively interacting with the latter), so often on the keyboard reorder panel button of another item type. This is seemingly just a VoiceOver/Chrome issue? I did my best to demo it here, but it's happening a bit inconsistently (at times focus appears stuck but the announcements match up to where native focus is moving/how user interactions respond, but at other times, the screen reader-imposed focus does truly remain stuck behind, and interactions impact the control outlined by VO:
  
  
    https://github.com/user-attachments/assets/2c6f061c-4a2c-4966-b3bd-14b0bd330154
  
    Video description: Tabbing through Item Types edit page. Drawers are being moved up and down via keyboard reorder panels. During these interactions, focus outline from VoiceOver keeps separating from native focus, attaching to controls above or below.
  
  - Also, not really getting consistent directional readouts on VoiceOver/all browsers (often just getting “Moved above” or “Moved below”, or otherwise just the name of the respective node)
  - When a new element is added, should focus move to the radio buttons instead of the “Add Element” button? I wonder if this more closely follows the logic of the change made for new nav items.
  - A bit outside this issue, but draggable handles are missing from newly added items.
- **Item > Files:**
  - VoiceOver, NVDA, JAWS on all browsers: Not getting any directional info, not even “Moved above” or “Moved below”
  - A bit outside this issue, but there’s an additional, non-visible and seemingly nonfunctional spot where focus is being brought after tabbing through all drawer content: a `.delete-checkbox` input.

Did the keyboard reorder buttons / panel previously have a description linking back to the item name, as expand/collapse and delete/undo buttons nearby often do? Perhaps I'm misremembering that they did, but in any case, I would think they should also announce/link to the overarching item.

**Comment by kimisgold** [comment]
> This seems like a good solution. Can the right margin be cut down at all? It feels like it's cutting them off with quite a bit of space left to go, but perhaps there's a practical reason for this.

I was making room for the expanded keyboard reordering panel, but realized that flexbox can do this width management for me just fine. Titles should now take up the full width.

> Focus keeps appearing to become stuck on buttons right above or below move up/move down buttons (while solely and consecutively interacting with the latter), so often on the keyboard reorder panel button of another item type. This is seemingly just a VoiceOver/Chrome issue?

I think VoiceOver and the browsers struggle to sync the screen reader cursor and focus cursor. I'm not sure there's anything we can do about that. I've seen it happen in screen reader tests from w3c.

I've added more descriptive labeling for the buttons so that the group label is announced with them. I've also simplified how the alerts are constructed so that we get more consistency across the browsers. I've focused on Appearance > Navigation for the moment before I do more changes across the other interfaces, tested with VoiceOver, JAWS, and NVDA. At the moment, I'm only getting inconsistent results with Safari for reasons I haven't been able to yet discern. If I can get the reordering announcements for the navigation across all the other browsers, that's still a big win. Could you look at the updates there?

**Comment by aviannamiller** [comment]
Updated title widths look great.

So just looking at Appearance > Navigation, I'm generally getting better results, especially with VoiceOver. I'm actually not experiencing inconsistencies with Safari, which is interesting. As was the case previously, I am still not getting reorder status announcements at all for JAWS on Firefox, just the button labels.

NVDA buttons have repetitive readouts of the drawer title on Edge and Firefox, but not Chrome, oddly. This appears to be the only screen reader that does this.

For VoiceOver and JAWS, all browsers, further "Cannot reorder further" alerts are not present for consecutive impossible actions, just the first attempted action (for instance, trying to move up the top-most link presents the alert, but then trying to nest that same link will not give any sort of alert beyond the control label). This was an issue noticed earlier in testing, as well. NVDA is only one that consistently alerts.

A new issue seems to have presented, seemingly some mismatch in which labels are linking to which items via `id` and `aria-labelledby`. This looks to be caused by a number of instances of shared `id`s. Here's an example of two drawers with the same `id`, causing an improper label readout on one of them:
<img width="1280" height="565" alt="Navigation appearance interface, displaying a number of moveable drawers. Keyboard focus is on a keyboard reorder panel button for a drawer labeled as 'Browse' in the interface. However, on-screen VoiceOver readout reveals the title as 'More information'." src="https://github.com/user-attachments/assets/37bad668-e9da-4c5b-b471-cb2ddaa0d666" />
<img width="1255" height="550" alt="Same interface as above. Browser dev tools are open, selecting keyboard reorder panel button for a drawer labeled as 'Browse' in the interface. An aria-labelledby attribute in the html reveals the label to be pulling from 'drawer-21', an id assigned to the same drawer label." src="https://github.com/user-attachments/assets/f0b33fcd-f437-4db7-94cd-a8f29c43d53d" />
<img width="1255" height="554" alt="Same interface as above. Browser dev tools are open, this time with the label for the actual 'More information' drawer selected. This drawer has the same id, 'drawer-21', as the 'Browse' drawer." src="https://github.com/user-attachments/assets/3171e621-2e75-47cb-af43-d987449cf674" />

**Comment by kimisgold** [comment]
* I've now provided a fix so that alerts consistently fire when the user attempts invalid reorder actions. This should work across the browsers.
* I've just tested JAWS 2026 with Firefox 153.0.1 on Windows, and I'm getting announcements from buttons and reorder actions.
* The repetitive button announcements on NVDA are coming from the `title` attributes. I'm tempted to leave it for now, and address it in the tooltip feature.
* I'm having trouble recreating the repeated IDs on the nav items. Could you provide a live example I could look at?
